stdenv-setup: Add quotes that don't do anything for consistency.

Ericson2314 opened the pull request:
###### Motivation for this change

@vcunat and others rightly point out that it's easier to quote always than learn Bash's idiosyncrasies enough to know when it doesn't make a difference.

This reverts commit 2743078f664ae07c4bed06a96182c6a86bd7fa32, which removes quotes that don't do anything, and then goes further adding even more quotes.

###### Things done

Please check what applies. Note that these are not hard requirements but mereley serve as information for reviewers.

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built stdenv on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
stdenv-setup: Add quotes that don't do anything for consistency.

@vcunat and others rightly point out that it's easier to quote always,
than learn Bash's idiosyncrasies enough to know when it doesn't make a
difference.

This reverts commit 2743078f664ae07c4bed06a96182c6a86bd7fa32, which
removes quotes that don't do anything, and then goes further adding
even more quotes.

copumpkin commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -141,11 +141,11 @@ trap "exitHandler" EXIT
+    local dir="$3"
     if [ -d "$dir" ]; then
-        eval export ${varName}=${!varName}${!varName:+$delimiter}${dir}
+        export "${varName}=${!varName}${!varName:+$delimiter}${dir}"
```
OMG NONTRIVIAL CHANGE

Ericson2314 replied:
Oh. I had this branch sitting around for a while before my current tail-between-legs conservationism :D. But, spare me this one eval removal if stdenv builds :D:D?

copumpkin commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -182,31 +182,31 @@ _addRpathPrefix() {
     local magic
     exec {fd}< "$fn"
-    read -r -n 4 -u $fd magic
+    read -r -n 4 -u "$fd" magic
```
I don't even know what that `exec {fd}< "$fn"` means. Is `fd` guaranteed to have no spaces here?

Ericson2314 replied:
Yes, it is.

edolstra replied:
It's a file descriptor, so yes. From the bash manpage:

> Each redirection that may be preceded by a file descriptor number may instead be preceded by a word of the form {varname}.  In this case, for each redirection operator except >&- and <&-, the shell will allocate a file  descriptor  greater than or equal to 10 and assign it to varname.

copumpkin commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -182,31 +182,31 @@ _addRpathPrefix() {
     if ! [ -x /bin/sh ]; then return 0; fi
     exec {fd}< "$fn"
-    read -r -n 2 -u $fd magic
+    read -r -n 2 -u "$fd" magic
```
Same question as above

Ericson2314 replied:
Even before my general PR, my basic is philosophy with bash is

- `"$asdf"` is a string substitution
- `"${asdf[@]}"` is an array string substitution
- `$asdf` is only for intentional re-splitting, and must be paired with a comment

In other words, don't forgo quoting because of a lack of deliminator in the definition. Even if that is obviously true it is still tempting fate, and per my bullets above is misleading because there is no intention of a word-splitting.

copumpkin reviewed (approved):
Looks pretty safe to me, but I'm not really a bash guru

Ericson2314 committed:
darwin stdenv: Ensure libSystem reexports the right libraries

The logic was made pure for the normal libSystem, but this change never
made it to the bootstrap tools. Deduplication the logic as the comment
suggests would have prevented this, but here's a stop-gap until we do
so.

Ericson2314 commented:
Built on both repos. I can now commence with the master -> staging merge :).
